This project uses a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The aim of this project is to clean the data and analyze the included used car listings. 

In [1]:
import pandas as pd
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns = ['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen']

In [6]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuelType  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [7]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  registration_year  gearbox       powerPS  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuelType       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [8]:
nrOfPictures = autos["nrOfPictures"]
nrOfPictures_counts = nrOfPictures.value_counts()
nrOfPictures_counts

0    50000
Name: nrOfPictures, dtype: int64

In [9]:
seller = autos["seller"]
seller_counts = seller.value_counts()
seller_counts

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [10]:
offer_type = autos["offerType"]
offer_type_counts = offer_type.value_counts()
offer_type_counts

Angebot    49999
Gesuch         1
Name: offerType, dtype: int64

- nrOfPictures has one value and is  candidate to be dropped
- Seller and offerType have only two different values.
- Price and odometer columns are numeric values stored as text
- We need to remove any non-numeric characters.
- Convert the column to a numeric dtype.

In [11]:
price = autos["price"]
price_counts = price.value_counts()
price_counts

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$37,450       1
$3,580        1
$7,420        1
$139          1
$19,777       1
Name: price, Length: 2357, dtype: int64

In [12]:
autos["price"] = autos["price"].str.replace('$','')
autos["price"] = autos["price"].str.replace(',','')
autos["price"] = autos["price"].astype(int)


In [13]:
price = autos["price"]
price_counts = price.value_counts()
price_counts

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

In [14]:
autos.rename(columns = {'odometer':'odometer_km'}, inplace = True) 

In [15]:
odometer = autos["odometer_km"]
odometer_counts = odometer.value_counts()
odometer_counts

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer_km, dtype: int64

In [16]:
autos["odometer_km"] = autos["odometer_km"].str.replace('km','')
autos["odometer_km"] = autos["odometer_km"].str.replace(',','')
autos["odometer_km"] = autos["odometer_km"].astype(int)

In [17]:
odometer = autos["odometer_km"]
odometer_counts = odometer.value_counts()
odometer_counts

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [18]:

autos = autos.drop(["nrOfPictures", "seller", "offerType"], axis=1)


In [19]:
autos["odometer_km"].unique().shape


(13,)

In [20]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [21]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

There are 13 unique rows of data for the odometer_km column.  The smalles reading is 5000km with 967 ocurrences. The mayority of the values are on the high end of 150000km. The values seem to be realistic.

In [22]:
autos["price"].unique().shape

(2357,)

In [23]:
autos["price"].describe().astype(int).astype(str)

count       50000
mean         9840
std        481104
min             0
25%          1100
50%          2950
75%          7200
max      99999999
Name: price, dtype: object

In [24]:
autos["price"].value_counts().sort_index(ascending=True)

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

In [25]:
autos["price"].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price, Length: 2357, dtype: int64

There are 1421 rows of cars with a price of $0 which seems odd.  This value makes up around 2.8\% of the data so we should be able to safely remove these rows of data. The max value is of 99999999 which seems odd, we need to investigate further.

We remove the outliers, so is not normal to have a price of 0, as bids on Ebay start at 1, also there is a big jump between      27322222 and 99999999, so we will remove the 99999999  value

In [26]:
autos = autos[autos['price'].between(1, 27322222)]

In [27]:
autos["price"].describe().astype(int).astype(str)

count       48578
mean         8069
std        180035
min             1
25%          1200
50%          3000
75%          7490
max      27322222
Name: price, dtype: object

Exploring the date columns

In [28]:
autos[['dateCrawled','ad_created','lastSeen']][0:5]

dateCrawled           ad_created             lastSeen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [29]:
(autos['dateCrawled']
        .str[:10]
        .value_counts(normalize = True, dropna = False)
        .sort_index()
        )

2016-03-05    0.025320
2016-03-06    0.014039
2016-03-07    0.036004
2016-03-08    0.033328
2016-03-09    0.033101
2016-03-10    0.032175
2016-03-11    0.032566
2016-03-12    0.036930
2016-03-13    0.015666
2016-03-14    0.036539
2016-03-15    0.034275
2016-03-16    0.029602
2016-03-17    0.031640
2016-03-18    0.012907
2016-03-19    0.034769
2016-03-20    0.037877
2016-03-21    0.037404
2016-03-22    0.032998
2016-03-23    0.032216
2016-03-24    0.029334
2016-03-25    0.031599
2016-03-26    0.032196
2016-03-27    0.031084
2016-03-28    0.034851
2016-03-29    0.034131
2016-03-30    0.033678
2016-03-31    0.031846
2016-04-01    0.033678
2016-04-02    0.035469
2016-04-03    0.038598
2016-04-04    0.036519
2016-04-05    0.013092
2016-04-06    0.003170
2016-04-07    0.001400
Name: dateCrawled, dtype: float64

We can notice there is uniformity in the distribution of the listings crawled each day.


In [30]:
(autos['ad_created']
        .str[:10]
        .value_counts(normalize = True, dropna = False)
        .sort_index()
        )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038845
2016-04-04    0.036889
2016-04-05    0.011816
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

We can see that there are more ads creating between march and april of 2016

In [31]:
(autos['lastSeen']
        .str[:10]
        .value_counts(normalize = True, dropna = False)
        .sort_index()
        )

2016-03-05    0.001070
2016-03-06    0.004323
2016-03-07    0.005393
2016-03-08    0.007431
2016-03-09    0.009613
2016-03-10    0.010663
2016-03-11    0.012372
2016-03-12    0.023797
2016-03-13    0.008893
2016-03-14    0.012598
2016-03-15    0.015871
2016-03-16    0.016448
2016-03-17    0.028079
2016-03-18    0.007349
2016-03-19    0.015830
2016-03-20    0.020647
2016-03-21    0.020647
2016-03-22    0.021368
2016-03-23    0.018527
2016-03-24    0.019762
2016-03-25    0.019206
2016-03-26    0.016798
2016-03-27    0.015645
2016-03-28    0.020874
2016-03-29    0.022356
2016-03-30    0.024764
2016-03-31    0.023797
2016-04-01    0.022788
2016-04-02    0.024929
2016-04-03    0.025197
2016-04-04    0.024476
2016-04-05    0.124768
2016-04-06    0.221808
2016-04-07    0.131912
Name: lastSeen, dtype: float64

We can see that mayority of distributions are wthin abril 5 until april 7,  most likely due to a sale. If the car has been sold, the ad is delisted from the site.

In [32]:
autos["registration_year"].describe()


count    48578.000000
mean      2004.753119
std         88.632571
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

As we can see above there are rows in the registration_year column that are incorrect such as the minimum value of 1000, before cars were invented, and a maximum value of 9999, many years in the future.

We can remove any rows with registration years above 2016 as this data was obtained in 2016. To be on the safe side we can see what percentage of vehicles have registration years between 1900 to 2016 and determine if it is a significant number.

In [33]:
autos['registration_year'].between(1900,2016).sum() / autos.shape[0]


0.9611758409156408

As we can see less that 4\% of vehicles are outside that range, so we can remove anything that falls outside it.

In [34]:
autos = autos[autos['registration_year'].between(1900, 2016)]

In [35]:
autos['registration_year'].describe()

count    46692.000000
mean      2002.908250
std          7.190078
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [36]:
autos['registration_year'].value_counts(normalize=True)

2000    0.067592
2005    0.062880
1999    0.062066
2004    0.057890
2003    0.057804
          ...   
1938    0.000021
1948    0.000021
1927    0.000021
1931    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

In [37]:
brands = autos["brand"].value_counts(normalize=True)
brands

volkswagen        0.211278
bmw               0.110040
opel              0.107577
mercedes_benz     0.096441
audi              0.086546
ford              0.069926
renault           0.047139
peugeot           0.029834
fiat              0.025636
seat              0.018269
skoda             0.016405
nissan            0.015270
mazda             0.015185
smart             0.014157
citroen           0.014028
toyota            0.012700
hyundai           0.010023
sonstige_autos    0.009873
volvo             0.009145
mini              0.008760
mitsubishi        0.008224
honda             0.007839
kia               0.007068
alfa_romeo        0.006639
porsche           0.006125
suzuki            0.005932
chevrolet         0.005697
chrysler          0.003512
dacia             0.002634
daihatsu          0.002506
jeep              0.002270
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001563
daewoo            0.001499
trabant           0.001392
r

In [38]:
common_brands = brands[brands > 0.05].index
common_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [39]:
mean_price_brand = {}
for brand in common_brands:
    current_brand =  autos[autos['brand'] == brand]
    mean_price = current_brand['price'].mean()
    mean_price_brand[brand] = mean_price

mean_price_brand

{'volkswagen': 6729.81956411556,
 'bmw': 8571.480147917478,
 'opel': 5432.479195699781,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 7456.547932618683}

We can see  Audi, BMW and Mercedes Benz cars are more expesive  than the competition, while Ford and Opel are less expensive

Convert both dictionaries to series objects, using the series constructor

In [40]:
mean_price_series = pd.Series(mean_price_brand).sort_values(ascending = False)
mean_price_series

audi             9336.687454
mercedes_benz    8628.450366
bmw              8571.480148
ford             7456.547933
volkswagen       6729.819564
opel             5432.479196
dtype: float64

In [41]:
mean_mileage_brand = {}
for brand in common_brands:
    current_brand =  autos[autos['brand'] == brand]
    mean_mileage = current_brand['odometer_km'].mean()
    mean_mileage_brand[brand] = int(mean_mileage)

mean_mileage_brand

{'volkswagen': 128713,
 'bmw': 132575,
 'opel': 129314,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124243}

Convert  to series objects, using the series constructor

In [42]:
mean_mileage_series = pd.Series(mean_mileage_brand).sort_values(ascending = False)
mean_mileage_series

bmw              132575
mercedes_benz    130788
opel             129314
audi             129157
volkswagen       128713
ford             124243
dtype: int64

In [43]:
top_brand_price = pd.DataFrame(mean_mileage_series,columns=['mean_mileage_series'])
top_brand_price

mean_mileage_series
bmw                         132575
mercedes_benz               130788
opel                        129314
audi                        129157
volkswagen                  128713
ford                        124243

Identify categorical data that uses german words, translate them and map the values to their english counterparts


In [44]:
autos['gearbox'].value_counts()


manuell      34722
automatik     9857
Name: gearbox, dtype: int64

In [45]:
autos['unrepaired_damage'].value_counts()


nein    33840
ja       4541
Name: unrepaired_damage, dtype: int64

In [46]:
autos['fuelType'].value_counts()


benzin     28546
diesel     14034
lpg          649
cng           71
hybrid        37
elektro       19
andere        15
Name: fuelType, dtype: int64

In [47]:
autos['vehicleType'].value_counts()


limousine     12602
kleinwagen    10585
kombi          8932
bus            4032
cabrio         3016
coupe          2465
suv            1965
andere          390
Name: vehicleType, dtype: int64

In [48]:
gearbox = {
    'manuell': 'manual',
    'automatik': 'automatic'
}

autos['gearbox'] = autos['gearbox'].map(gearbox)


In [49]:
unrepaired_damage = {
    'nein': 'no',
    'ja': 'yes'
}
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(unrepaired_damage)


In [50]:
autos['fuelType'] = autos['fuelType'].replace(['benzin', 'elektro', 'andere'], ['gasoline', 'electric', 'other'])


In [51]:
autos['vehicleType'] = autos['vehicleType'].replace(['limousine', 'kleinwagen', 'kombi', 'cabrio', 'andere'], ['sedan', 'small car', 'stationwagen', 'convertible', 'other'])


In [52]:
autos['model'] = autos['model'].replace('andere', 'other')


In [53]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest   vehicleType  registration_year    gearbox  powerPS  \
0   5000  control           bus               2004     manual      158   
1   8500  control         sedan               1997  automatic      286   
2   8990     test         sedan               2009     manual      102   
3   4350  control     small car               2007  automatic       71   
4   1350     test  stationwagen               2003     manual        0   

    model  odometer_km  registration_month  fuelType       brand  \
0   other       150000                   3       lpg     peugeot   
1     7er       150000                   6  gasoline         bmw   
2    golf        70000                   7  gasoline  volkswagen   
3  fortwo        70000                   6  gasoline       smart   
4   focus       150000                   7  gasoline        ford   

  unrepaired_damage           ad_created  postalCode             lastSeen  
0                no  2016-03-26 00:00:00       79588  2016-04-06 06:45:54  
1                no  2016-04-04 00:00:00       71034  2016-04-06 14:45:08  
2                no  2016-03-26 00:00:00       35394  2016-04-06 20:15:37  
3                no  2016-03-12 00:00:00       33729  2016-03-15 03:16:28  
4                no  2016-04-01 00:00:00       39218  2016-04-01 14:38:50

How much cheaper are cars with damage than their non-damaged counterparts?


In [54]:
no_yes = ['no', 'yes']

damaged_prices = {}

for isdamaged in no_yes:
    damage = autos[autos['unrepaired_damage'] == isdamaged]
    mean_price = damage['price'].mean()
    damaged_prices[isdamaged] = int(mean_price)
    
damaged_prices

{'no': 8334, 'yes': 2460}

We can see that cars damaged cars are considerable cheaper than those with no damage prior to their listing.

In [55]:
just_brands = brands = autos["brand"].value_counts(normalize=True).index
just_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai', 'sonstige_autos', 'volvo', 'mini',
       'mitsubishi', 'honda', 'kia', 'alfa_romeo', 'porsche', 'suzuki',
       'chevrolet', 'chrysler', 'dacia', 'daihatsu', 'jeep', 'subaru',
       'land_rover', 'saab', 'jaguar', 'daewoo', 'trabant', 'rover', 'lancia',
       'lada'],
      dtype='object')

In [56]:

freq_comb = autos.groupby(['brand','model']).size().reset_index()
freq_comb.columns = ['brand','model','count']
freq_comb.sort_values(['count'],ascending=False)


brand       model  count
266  volkswagen        golf   3707
23          bmw         3er   2615
273  volkswagen        polo   1609
181        opel       corsa   1592
271  volkswagen      passat   1349
..          ...         ...    ...
212       rover   discovery      1
215       rover  rangerover      1
77         ford       b_max      1
27          bmw          i3      1
7          audi         200      1

[290 rows x 3 columns]

We can see that the volkswagen golf is the most common brand/model combination

In [57]:
autos["odometer_km"].value_counts()

150000    30090
125000     4858
100000     2059
90000      1673
80000      1375
70000      1187
60000      1128
50000       995
40000       798
5000        786
30000       760
20000       742
10000       241
Name: odometer_km, dtype: int64

In [58]:
odometer_km_g1 = autos[autos["odometer_km"].between(241, 40000)]
odometer_km_g2 = autos[autos["odometer_km"].between(40001, 80000)]
odometer_km_g3 = autos[autos["odometer_km"].between(80001, 150000)]

values1 = odometer_km_g1.groupby(['odometer_km','price']).size().reset_index()
values1.columns = ['odometer_km','price','count']
values1.sort_values(by=['odometer_km'],ascending=False).head(10)


odometer_km     price  count
1907        40000  27322222      1
1615        40000      8650      1
1604        40000      8199      1
1605        40000      8200      3
1606        40000      8250      1
1607        40000      8299      3
1608        40000      8300      2
1609        40000      8400      3
1610        40000      8490      2
1611        40000      8495      1

In [59]:
values2 = odometer_km_g2.groupby(['odometer_km','price']).size().reset_index()
values2.columns = ['odometer_km','price','count']
values2.sort_values(by=['odometer_km'],ascending=False).head(10)


odometer_km   price  count
2152        80000  105000      1
1751        80000    5380      1
1767        80000    5924      1
1766        80000    5900      5
1765        80000    5890      1
1764        80000    5888      1
1763        80000    5879      1
1762        80000    5800      2
1761        80000    5799      1
1760        80000    5750      2

In [60]:
values3 = odometer_km_g3.groupby(['odometer_km','price']).size().reset_index()
values3.columns = ['odometer_km','price','count']
values3.sort_values(by=['odometer_km'],ascending=False).head(10)


odometer_km     price  count
3575       150000  12345678      1
2595       150000      3020      1
2586       150000      2990     92
2587       150000      2993      1
2588       150000      2995      4
2589       150000      2997      1
2590       150000      2998      1
2591       150000      2999    160
2592       150000      3000    257
2593       150000      3001      1

The price gets lower as the odometer_km gets bigger in most cases